File: server/src/models/request.ts
import { Model, DataTypes, Optional } from "sequelize";
import { sequelize } from "@config/database";

export enum RequestType {
	// Standard Leaves
	LEAVE = "LEAVE", // General/Annual leave
	SICK = "SICK", // Sick leave
	UNPAID = "UNPAID", // Unpaid leave

	// Time & Attendance Specifics
	LATE_EARLY = "LATE_EARLY", // Late arrival / early leave
	OVERTIME = "OVERTIME", // Overtime request (OT)
	BUSINESS_TRIP = "BUSINESS_TRIP", // Business trip
	SHIFT_CHANGE = "SHIFT_CHANGE", // Shift change request
	REMOTE_WORK = "REMOTE_WORK", // Remote work request

	// Corrections & Admin
	ATTENDANCE_CONFIRMATION = "ATTENDANCE_CONFIRMATION",
	ATTENDANCE_ADJUSTMENT = "ATTENDANCE_ADJUSTMENT",
	EXPLANATION = "EXPLANATION", // Explaining a missing check-in

	// Fallback
	OTHER = "OTHER",
}

export const RequestTypeLabels: Record<RequestType, string> = {
	[RequestType.LEAVE]: "Leave Request",
	[RequestType.SICK]: "Sick Leave",
	[RequestType.UNPAID]: "Unpaid Leave",
	[RequestType.LATE_EARLY]: "Late Arrival / Early Leave",
	[RequestType.OVERTIME]: "Overtime (OT)",
	[RequestType.BUSINESS_TRIP]: "Business Trip",
	[RequestType.SHIFT_CHANGE]: "Shift Change",
	[RequestType.REMOTE_WORK]: "Remote Work",
	[RequestType.ATTENDANCE_CONFIRMATION]: "Attendance Confirmation",
	[RequestType.ATTENDANCE_ADJUSTMENT]: "Attendance Adjustment",
	[RequestType.EXPLANATION]: "Explanation",
	[RequestType.OTHER]: "Other",
};

export enum RequestStatus {
	PENDING = "PENDING",
	APPROVED = "APPROVED",
	REJECTED = "REJECTED",
}

export const RequestStatusLabels: Record<RequestStatus, string> = {
	[RequestStatus.APPROVED]: "Approved",
	[RequestStatus.PENDING]: "Pending",
	[RequestStatus.REJECTED]: "Rejected",
};

interface RequestAttributes {
	id: string;
	userId: string;
	type: RequestType;
	fromDate?: Date | null;
	toDate?: Date | null;
	reason: string;
	attachments?: string | null; // JSON array of file paths
	status: RequestStatus;
	reviewedBy?: string | null;
	reviewNote?: string | null;
}

interface RequestCreationAttributes
	extends Optional<
		RequestAttributes,
		"id" | "status" | "attachments" | "reviewedBy" | "reviewNote" | "fromDate" | "toDate"
	> {}

export default class RequestModel
	extends Model<RequestAttributes, RequestCreationAttributes>
	implements RequestAttributes
{
	public declare id: string;
	public declare userId: string;
	public declare type: RequestType;
	public declare fromDate: Date | null;
	public declare toDate: Date | null;
	public declare reason: string;
	public declare attachments: string | null;
	public declare status: RequestStatus;
	public declare reviewedBy: string | null;
	public declare reviewNote: string | null;

	public declare readonly createdAt: Date;
	public declare readonly updatedAt: Date;
}

RequestModel.init(
	{
		id: {
			type: DataTypes.UUID,
			defaultValue: DataTypes.UUIDV4,
			primaryKey: true,
		},
		userId: {
			type: DataTypes.UUID,
			allowNull: false,
		},
		type: {
			type: DataTypes.ENUM(...Object.values(RequestType)),
			allowNull: false,
			defaultValue: RequestType.OTHER,
		},
		fromDate: {
			type: DataTypes.DATE,
			allowNull: true,
		},
		toDate: {
			type: DataTypes.DATE,
			allowNull: true,
		},
		reason: {
			type: DataTypes.TEXT,
			allowNull: false,
		},
		attachments: {
			type: DataTypes.TEXT, // JSON array of file paths
			allowNull: true,
		},
		status: {
			type: DataTypes.ENUM(...Object.values(RequestStatus)),
			allowNull: false,
			defaultValue: RequestStatus.PENDING,
		},
		reviewedBy: {
			type: DataTypes.UUID,
			allowNull: true,
		},
		reviewNote: {
			type: DataTypes.TEXT,
			allowNull: true,
		},
	},
	{
		sequelize,
		tableName: "requests",
		underscored: true,
		timestamps: true,
	}
);

File: client/lib/src/models/request.dart
import 'dart:convert';

enum RequestType {
  LEAVE,
  SICK,
  UNPAID,
  LATE_EARLY,
  OVERTIME,
  BUSINESS_TRIP,
  SHIFT_CHANGE,
  REMOTE_WORK,
  ATTENDANCE_CONFIRMATION,
  ATTENDANCE_ADJUSTMENT,
  EXPLANATION,
  OTHER;

  static RequestType fromString(String? value) {
    if (value == null || value.isEmpty) return RequestType.OTHER;
    final v = value.toUpperCase();
    return RequestType.values.firstWhere(
      (e) => e.name == v,
      orElse: () => RequestType.OTHER,
    );
  }

  String toTextString() {
    final s = name.replaceAll('_', ' ').toLowerCase();
    return s.isNotEmpty ? '${s[0].toUpperCase()}${s.substring(1)}' : '';
  }
}

enum RequestStatus {
  PENDING,
  APPROVED,
  REJECTED;

  static RequestStatus fromString(String? value) {
    if (value == null || value.isEmpty) return RequestStatus.PENDING;
    final v = value.toUpperCase();
    return RequestStatus.values.firstWhere(
      (e) => e.name == v,
      orElse: () => RequestStatus.PENDING,
    );
  }
}

class Request {
  final String? id;
  final String userId;
  final RequestType type;
  final DateTime? fromDate;
  final DateTime? toDate;
  final String? reason;
  final List<String>? attachments;
  final RequestStatus status;
  final String? reviewedBy;
  final String? reviewNote;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? userName;

  Request({
    this.id,
    required this.userId,
    required this.type,
    this.fromDate,
    this.toDate,
    this.reason,
    this.attachments,
    this.status = RequestStatus.PENDING,
    this.reviewedBy,
    this.reviewNote,
    this.createdAt,
    this.updatedAt,
    this.userName,
  });

  factory Request.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic v) {
      if (v == null) return null;
      try {
        return DateTime.parse(v.toString());
      } catch (_) {
        return null;
      }
    }

    List<String>? parseAttachments(dynamic v) {
      if (v == null) return null;
      if (v is List) return v.map((e) => e.toString()).toList();
      if (v is String) {
        try {
          final decoded = jsonDecode(v);
          if (decoded is List) return decoded.map((e) => e.toString()).toList();
        } catch (_) {}
      }
      return null;
    }

    return Request(
      id: json['id'],
      userId: json['userId'],
      type: RequestType.fromString(json['type']),
      fromDate: json['fromDate'] != null
          ? DateTime.parse(json['fromDate'])
          : null,
      toDate: json['toDate'] != null ? DateTime.parse(json['toDate']) : null,
      reason: json['reason'],
      attachments: parseAttachments(json['attachments']),
      status: RequestStatus.fromString(json['status']),
      reviewedBy: json['reviewedBy'],
      reviewNote: json['reviewNote'],
      createdAt: parseDate(json['createdAt']),
      updatedAt: parseDate(json['updatedAt']),
      userName: json['user'] != null ? json['user']['name'] : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'type': type.name,
      'from_date': fromDate?.toIso8601String(),
      'to_date': toDate?.toIso8601String(),
      'reason': reason,
      'attachments': attachments != null ? jsonEncode(attachments) : null,
      'reviewed_by': reviewedBy,
      'review_note': reviewNote,
      'status': status.name,
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }
}
